src/frontend/Ast_to_Mir.ml

```diff
@@ -299,41 +299,61 @@ let extract_transform_args var = function
    |TupleTransformation _ | StochasticRow | StochasticColumn ->
       []
 
-let rec param_size transform sizedtype =
-  let rec shrink_eigen f st =
-    match st with
-    | SizedType.SArray (t, d) -> SizedType.SArray (shrink_eigen f t, d)
-    | SVector (mem_pattern, d) | SMatrix (mem_pattern, d, _) ->
-        SVector (mem_pattern, f d)
-    | SInt | SReal | SComplex | SRowVector _ | STuple _ | SComplexRowVector _
-     |SComplexVector _ | SComplexMatrix _ ->
+(** Allows [shrink_helper] to operate either on each dimension independently
+   or on both matrix dimensions at once *)
+type size_change =
+  | Univariate of (Expr.Typed.t -> Expr.Typed.t)
+  | Multivariate of (Expr.Typed.t -> Expr.Typed.t -> Expr.Typed.t)
+
+(** We need to compute somewhat arbitrary new sizes for the unconstrained
+    parameters. This function handles the primary cases:
+
+  - A vector of size N is transformed to a vector of size (f N)
+  - A matrix of size N x M is transformed to a vector of size (f N) x (f_d2 M)
+  - A matrix of size N x N is transformed to a vector of size (f N M)
+  - Arrays of the above are handled recursively
+*)
+let rec shrink_helper (f : size_change) f_d2 st =
+  let f_assert_univariate d =
+    match f with
+    | Univariate f -> f d
+    | Multivariate _ ->
         Common.ICE.internal_compiler_error
           [%message
-            "Expecting SVector or SMatrix, got " (st : Expr.Typed.t SizedType.t)]
-  in
-  let rec shrink_eigen_mat f st =
-    match st with
-    | SizedType.SArray (t, d) -> SizedType.SArray (shrink_eigen_mat f t, d)
-    | SMatrix (mem_pattern, d1, d2) -> SVector (mem_pattern, f d1 d2)
-    | SInt | SReal | SComplex | SRowVector _ | SVector _ | STuple _
-     |SComplexRowVector _ | SComplexVector _ | SComplexMatrix _ ->
-        Common.ICE.internal_compiler_error
-          [%message "Expecting SMatrix, got " (st : Expr.Typed.t SizedType.t)]
-  in
+            "To shrink a vector, the first argument must be a univariate \
+             function "
+              (st : Expr.Typed.t SizedType.t)] in
+  match st with
+  | SizedType.SArray (t, d) -> SizedType.SArray (shrink_helper f f_d2 t, d)
+  | SVector (mem_pattern, d) -> SVector (mem_pattern, f_assert_univariate d)
+  | SRowVector (mem_pattern, d) ->
+      SRowVector (mem_pattern, f_assert_univariate d)
```

Comment:
Re: codecov
Do we not have row vector transforms?

Reply:
None that change the shape currently, e.g. there's nothing like a row_simplex